src/Microsoft.AspNetCore.Mvc.Razor/RazorPage.cs

```diff
@@ -316,18 +314,7 @@ public static void WriteTo(TextWriter writer, HtmlEncoder encoder, object value)
             var htmlContent = value as IHtmlContent;
             if (htmlContent != null)
             {
-                var htmlTextWriter = writer as HtmlTextWriter;
-                if (htmlTextWriter == null)
-                {
-                    htmlContent.WriteTo(writer, encoder);
-                }
-                else
-                {
-                    // This special case allows us to keep buffering as IHtmlContent until we get to the 'final'
-                    // TextWriter.
-                    htmlTextWriter.Write(htmlContent);
-                }
-
```

Comment:
Could also special case `value is IHtmlContentContainer && writer is IHtmlContentContainer`. Would that avoid the worst? If the performances are high in number, could revisit later -- as long as this PR leaves the correctnesses high.